### ISSUE TITLE
buffer per layer

### ISSUE THREAD

**Issue by andrewharvey** [mapbox/tippecanoe#385]
Would it be possible to set a `--buffer` per layer?

For context, I need to set a larger `--buffer` for a point/label layer to work around an issue where the Mapbox Raster Tiles from Styles API clips long labels short (GL JS doesn't seem to clip).

However this isn't needed for other polygon or line layers within the same tileset/mbtiles and makes the filesize and possibly tiles much larger than they need to be.

**Comment by ericfischer** [comment]
Hmm, yes, this should be possible, but like a lot of per-layer things, it requires coming up with a good syntax for talking about it.

The workaround at the moment would be to make the layers as separate mbtiles tiles, each with its own set of options, and then use `tile-join` to combine the separate mbtiles files into one.

**Comment by andrewharvey** [comment]
> The workaround at the moment would be to make the layers as separate mbtiles tiles, each with its own set of options, and then use tile-join to combine the separate mbtiles files into one.

I didn't realise tile-join could be used that way, works perfectly! With the workaround, per layer buffer in the one command is less important from my perspective

**Comment by ericfischer** [comment]
Great, I'm glad you were able to get `tile-join` to do what you needed!

**Comment by pathmapper** [comment]
Thanks, very helpful! 
There are cases where labels are also clipped in GL JS: https://github.com/mapbox/mapbox-gl-js/issues/2706

**Comment by andrewharvey** [comment]
> Thanks, very helpful!
> There are cases where labels are also clipped in GL JS: mapbox/mapbox-gl-js#2706

If you've set a large enough buffer in tippecanoe it should fix/workaround mapbox/mapbox-gl-js#2706

**Comment by pathmapper** [comment]
@andrewharvey you're right.
Do you have any ideas how to find the **smallest** large enough buffer for a label layer?
I 'd like to have the filesize as small as possible.
Currently I am using trial and error, but this is not productive having a layer with several hundreds labels.

**Comment by andrewharvey** [comment]
@pathmapper it depends how wide your widest label will be for the style used.

From the docs:

> -b pixels or --buffer=pixels: Buffer size where features are duplicated from adjacent tiles. Units are "screen pixels"—1/256th of the tile width or height. (default 5)

You might need to experiment to confirm, but I'm guessing a -b 256 will mean a label falling just outside a 256px tile X will render all the way across tile X if the label is big enough.

**Comment by pathmapper** [comment]
@andrewharvey yes, it depends on the widest label. But it's difficult to identify the width in pixels - sometimes there are zoom functions for the text-size and I think we must take all zoom levels in account for which the layer is used in the style.

Maybe it is possible to use [text-max-width](https://www.mapbox.com/mapbox-gl-js/style-spec/#layout-symbol-text-max-width) and limit and determine the label width that way.

Interesting how you describe buffers. My current understanding of the buffers is, a -b 256 for tile X will mean a label of tile X reaching across the tile border will be rendered up to 256px across the adjacent tile.
Here is more to read about buffers: [https://www.mapbox.com/help/studio-classic-sources/#data-layer-options](https://www.mapbox.com/help/studio-classic-sources/#data-layer-options)

**Comment by pathmapper** [comment]
@andrewharvey after I've read your description of buffers again, I think we are understanding buffers the same way:-)

**Comment by caneroj1** [comment]
If one of the main issues here is coming up with a syntax for how to talk about a per-layer buffer, can I suggest extending the existing syntax for specifying named layers?

According to the README, we have a few ways of specifying named layers:

- `-l` name or `--layer=`name
- `-L` name`:`file.json or `--named-layer=`name`:`file.json
- `-L{`layer-json`}` or `--named-layer={`layer-json`}`

Maybe something like the following would work?

- `-l` name`:`buffer or `--layer=`name`:`buffer
- `-L` name`:`file.json`:`buffer or `--named-layer=`name`:`file.json`:`buffer
- `-L{`layer-json`}` or `--named-layer={`layer-json`}`, with `buffer: x` as a property inside the layer-son.

I suppose if the buffer isn't specified for a layer, just fallback to the buffer setting for the mbtiles file as a whole.

**Comment by ericfischer** [comment]
The `-L{`*layer-json*`}` form with `buffer` as an additional property seems like the sustainable way to go forward with per-layer options. I started working on adding `buffer` as an option there but didn't get very far yet.
